plutus-metatheory/src/Untyped/Purity.lagda.md

```diff
@@ -0,0 +1,146 @@
+---
+title: VerifiedCompilation.Purity
+layout: page
+---
+
+# Definitions of Purity for Terms
+```
+module Untyped.Purity where
+
+```
+## Imports
+
+```
+open import Untyped using (_⊢; case; builtin; _·_; force; `; ƛ; delay; con; constr; error)
+open import Relation.Nullary using (Dec; yes; no; ¬_; _×-dec_)
+open import Builtin using (Builtin; arity)
+open import Utils as U using (Maybe;nothing;just)
+open import RawU using (TmCon)
+open import Data.Product using (_,_; _×_)
+open import Data.Nat using (ℕ; zero; suc; _>_; _>?_)
+open import Data.List using (List; _∷_; [])
+open import Data.List.Relation.Unary.All using (All)
+open import Data.Maybe using (Maybe; just; nothing; from-just)
+open import Data.Maybe.Properties using (just-injective)
+open import Agda.Builtin.Equality using (_≡_; refl)
+open import Relation.Nullary.Negation using (contradiction)
+open import Relation.Binary.PropositionalEquality.Core using (trans; _≢_; subst; sym; cong)
+open import Data.Empty using (⊥)
+open import Function.Base using (case_of_)
+open import Untyped.CEK using (lookup?)
+open import VerifiedCompilation.UntypedViews using (isDelay?; isTerm?; isdelay; isterm)
+-- FIXME: This moves to Untyped.Reduction in a different PR...
+open import VerifiedCompilation.UCaseReduce using (iterApp)
+```
+## Untyped Purity
+
+The definition here is based on the (Haskell Implementation)[https://github.com/IntersectMBO/plutus/blob/master/plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Purity.hs]. It is mostly uncontentious.
+```
+saturation : {X : Set} → X ⊢ → Maybe (ℕ × ℕ)
+saturation (builtin b) = just (((arity b), zero))
+saturation (t · _) with saturation t
+... | nothing = nothing
+... | just (arity , apps) = just (arity , suc apps)
+saturation (force t) with saturation t
+... | nothing = nothing
+... | just (arity , apps) = just (arity , suc apps)
+saturation _ = nothing
+
+sat-det : {X : Set} {arity args arity₁ args₁ : ℕ} {t t₁ : X ⊢}
+        → saturation t ≡ just (arity , args)
+        → saturation t₁ ≡ just (arity₁ , args₁)
+        → t ≡ t₁
+        → just (arity , args) ≡ just (arity₁ , args₁)
+sat-det sat-t sat-t₁ refl = trans (sym sat-t) sat-t₁
+
+data Pure {X : Set} : (X ⊢) → Set where
+    force : {t : X ⊢} → Pure t → Pure (force (delay t))
+
+    constr : {i : ℕ} {xs : List (X ⊢)} → All Pure xs → Pure (constr i xs)
+    -- case applied to constr would reduce, and possibly be pure.
+    case : {i : ℕ} {t : X ⊢}{vs ts : List (X ⊢)}
+           → lookup? i ts ≡ just t
+           → Pure (iterApp t vs)
+           → Pure (case (constr i vs) ts)
+    -- case applied to anything else is Unknown
+
+    builtin : {b : Builtin} → Pure (builtin b)
+
+    unsat-builtin : {t₁ t₂ : X ⊢} {arity args : ℕ}
+            → saturation t₁ ≡ just (arity , args)
+            → arity > (suc args)
```

Comment:
Still wrong as per the previous discussion. `Apply IfThenElse True` is impure, because it fails due to a missing `force`, but this logic here doesn't recognize that.

The implementation is being fixed.